File: solutions/search-a-2d-matrix/Solution.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int cols = matrix[0].length;
        int lo = 0, hi = matrix.length * cols - 1; // read row by row, it is one sorted list
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            int v = matrix[mid / cols][mid % cols]; // position k is row k / cols, column k % cols
            if (v == target) return true;
            if (v < target) lo = mid + 1;
            else hi = mid - 1;
        }
        return false;
    }
}

File: solutions/search-a-2d-matrix/solution.rs
use std::cmp::Ordering;

impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        let cols = matrix[0].len();
        let (mut lo, mut hi) = (0, matrix.len() * cols); // read row by row, one sorted list: [lo, hi)
        while lo < hi {
            let mid = lo + (hi - lo) / 2;
            match matrix[mid / cols][mid % cols].cmp(&target) {
                Ordering::Equal => return true,
                Ordering::Less => lo = mid + 1,
                Ordering::Greater => hi = mid,
            }
        }
        false
    }
}
